zkvm/src/program.rs

```diff
@@ -47,6 +47,26 @@ macro_rules! def_op {
     );
 }
 
+pub(crate) trait Encodable {
+   // Returns the serialized length of a datatype who implement this trait.
+   fn encode(&self, buf: &mut Vec<u8>);
+   /// Encodes a datatype who im0lements this trait into a buffer.
+   fn serialized_length(&self) -> usize;
```

Comment:
let's add another implemented method to factor out the common encode-to-vec pattern:

```rust
fn encode_to_vec(&self) -> Vec<u8> {
    let mut buf = Vec::with_capacity(self.serialized_length());
    self.encode(&mut buf);
    buf
}
```

Then, every time we do `buf=Vec::with_capacity(serialized_length) ... encode(&mut buf)` we can call that method instead.